### PR TITLE
WIP: QSSA encoding 

### PR DESCRIPTION
**Pull request by pingshiyu**
As an example dialect showcasing xDSL features which might be useful for other learning xDSL, and as a way for myself to get familiar with xDSL, I've been implementing the QSSA dialect (https://dl.acm.org/doi/pdf/10.1145/3497776.3517772) within xDSL as a first project.

It's still a WIP and not all of the features are faithfully implemented yet, however I'm opening this PR as a draft to get some feedback on the implementation so far. I'm still in learning stages of xDSL/MLIR so suggestions and better / cleaner ways to do things are very welcome!

A few issues I ran in to which I haven't been able to solve yet:
- Operation traits - are these in xDSL at the moment?
- Is it possible to implement dynamic tensors? i.e. tensors with dynamic types

### CODE REVIEW

**Comment by math-fehr** [comment]
Nice! This looks quite nice, I'll take a deeper look tomorrow if you want (mostly on the dialect definition, I'm not really familiar with the domain itself).

For Traits in xDSL, you can directly use class inheritance. Something like:
```
class MyOp(Operation, Pure):
   ...
```
This essentially gives traits and interfaces for free from Python. You can call the functions directly, and use `isinstance` to check if an operation satisfy a trait or not.

I don't think you can define dynamic tensors in MLIR, or if you could, this is definitely not implemented in xDSL. I'm not sure how this would work though, since when lowering to LLVM, we would need to know the element type size to be able to access elements in the tensor.

**Comment by pingshiyu** [comment]
Thank you very much Mathieu for your comments! It'd be very appreciate if you could take a closer look at the dialect definitions too, if possible!

Thank you also for your thoughts on implementing traits & dynamic arrays! I'll have more of a think about the dynamic arrays and will keep traits in mind for future :)

**Review by webmiche — COMMENTED** [review]
LGTM overall. I don't quite get why you are naming your builder `apply`, but I guess you are thinking of this differently than I do (maybe in terms of functons instead of operations?). Also, I would add some lit tests, I feel these are nicer to look at (at least that is how I think about IRs).

**Comment by pingshiyu** [comment]
@webmiche thank you for looking through and for your feedback! I will look into the lit tests :)

File: xdsl/dialects/qssa.py [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+        return Qubits([IntAttr.from_int(n_qubits)])
+
+
+q1_Type = Qubits.get_n_qubits_type(1)
+q2_Type = Qubits.get_n_qubits_type(2)
+
+
+# data attributes
+@irdl_attr_definition
+class Angle(Data[float]):
```
Comment (superlopuh) [inline xdsl/dialects/qssa.py:105]: what's the unit?

**Review by superlopuh — CHANGES_REQUESTED** [review]
Code mostly look good, marking as request changes until no longer WIP

**Comment by AntonLydike** [comment]
Is this PR still active?

**Comment by pingshiyu** [comment]
> 

not working on this anymore, will go ahead and close.

**Pull request closed without merging**